Implement some experimental HTTP server functionality ⭐

matiasinsaurralde opened the pull request:
Felt inspired by the last thing @ry mentioned in his talk: a ⭐️ for the first who implement a web server?

This is very experimental!

Sample server:
```
% deno -allow-net samples/server.ts 
```

Sample endpoints:

```
% curl http://localhost:5000/hello
Hello from Deno
% curl http://localhost:5000/echo -X POST --data '{"echo":"jsontest"}'
jsontest
```

- Any other endpoint will throw HTTP 400.
- It will try to parse request body as JSON by default (and this will be accessible from the server request callback). If this fails you will get a string representation of the body.
- You can `res.Write` strings and `Uint8Array`.
- Headers are still missing, think there's another PR that may help with this.
- Have added a bunch of PB stuff.
- The rps are pretty decent.
- No tests yet

Screen recording: https://youtu.be/vbFr29wMzSA

ry commented:
Very nice!
1. Have you done any benchmarking? How does this compare to Node?
2. The fetch API in the browser defines a [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) and [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) do you think these could be reused for servers?

ry commented:
⭐

matiasinsaurralde commented:
1. Yes, it was faster before extending the dispatcher logic. Now I've figured out a way to avoid handling an additional `map` here and I'm testing some refactoring around `Sub`. The basic idea is the current approach of the dispatcher and turn the handler code to something like this, every request would have its own channel:
```go
func buildHTTPHandler(serverID int32) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		// req := requestDispatcher.Store(r, w)

		// Increment and get an ID for this request:
		atomic.AddInt32(n, 1)
		reqID := *n

		// Used to signal end:
		done := make(chan bool)

		// Instead of flooding the "http" channel, use our own (might need a prefix later):
		ch := fmt.Sprintf("%d", reqID)

		// Subscribe to this channel and handle stuff:
		go Sub(fmt.Sprintf(ch), func(buf []byte) []byte {
			msg := &Msg{}
			proto.Unmarshal(buf, msg)
			switch msg.Command {
			case Msg_HTTP_RES_WRITE:
				w.Write(msg.HttpResBody)
			case Msg_HTTP_RES_STATUS:
				w.WriteHeader(int(msg.HttpResCode))
			case Msg_HTTP_RES_END:
				done <- true
			}
			return buf
		})

		// Prepare and publish request message:
		var body []byte
		if r.Body != nil {
			body, _ = ioutil.ReadAll(r.Body)
		}
		msg := &Msg{}
		msg.HttpReqId = reqID
		msg.HttpReqBody = body
		msg.Command = Msg_HTTP_REQ
		msg.HttpReqPath = r.URL.Path
		msg.HttpReqMethod = r.Method
		go PubMsg(httpChan, msg)

		w.Header().Set("Server", serverHeader)

		// We were using this before:
		// <-req.done
		// Now we block here (so no dispatcher overhead, no additional maps!)
		<-done
	}
}
```
1.1 Some benchmark output from a similar HTTP program with Node 8.6:
```
% boom -n 500 -c 20 http://localhost:5000/
Summary:
  Total:	0.0911 secs
  Slowest:	0.0179 secs
  Fastest:	0.0023 secs
  Average:	0.0036 secs
  Requests/sec:	5486.9475

Status code distribution:
  [200]	500 responses

Response time histogram:
  0.002 [1]	|
  0.004 [397]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.005 [66]	|∎∎∎∎∎∎
  0.007 [13]	|∎
  0.009 [3]	|
  0.010 [0]	|
  0.012 [0]	|
  0.013 [0]	|
  0.015 [2]	|
  0.016 [8]	|
  0.018 [10]	|∎

Latency distribution:
  10% in 0.0024 secs
  25% in 0.0026 secs
  50% in 0.0028 secs
  75% in 0.0033 secs
  90% in 0.0046 secs
  95% in 0.0056 secs
  99% in 0.0173 secs
```
1.2 This experiment with the new approach (not bad!):
```
% boom -n 500 -c 20 http://localhost:5000/
Summary:
  Total:	0.1146 secs
  Slowest:	0.0112 secs
  Fastest:	0.0034 secs
  Average:	0.0045 secs
  Requests/sec:	4362.9953

Status code distribution:
  [400]	500 responses

Response time histogram:
  0.003 [1]	|
  0.004 [291]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.005 [110]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.006 [54]	|∎∎∎∎∎∎∎
  0.007 [6]	|
  0.007 [3]	|
  0.008 [24]	|∎∎∎
  0.009 [4]	|
  0.010 [3]	|
  0.010 [0]	|
  0.011 [4]	|

Latency distribution:
  10% in 0.0036 secs
  25% in 0.0038 secs
  50% in 0.0041 secs
  75% in 0.0048 secs
  90% in 0.0057 secs
  95% in 0.0075 secs
  99% in 0.0094 secs
```
2. Yes, I did this in a separate way to have some flexibility but agree they should be shared somehow.

ry commented on http.ts:
```diff
@@ -0,0 +1,130 @@
+    method: string;
+    body: any | string;
+    constructor(url: string, opts?: RequestOptions) {
+      if(opts == null) {
```
style: there should be a space after the "if". Running "yarn lint --fix" should fix this, I think.

ry commented on http.ts:
```diff
@@ -0,0 +1,130 @@
+
+export class Request {
+    opts: RequestOptions;
+    readonly defaultMethod = "GET";
```
This member seems unnecessary.

ry commented on http.go:
```diff
@@ -0,0 +1,112 @@
+	})
+}
+
+var reqID = new(int32)
```
s/reqID/nextReqId/

Why is it a pointer? I think this can be just 
```
var nextReqId int32 = 0
```

ry commented on http.go:
```diff
@@ -0,0 +1,112 @@
+func getReqID() (int32, string) {
+	atomic.AddInt32(reqID, 1)
+	id := *reqID
+	return id, fmt.Sprintf("%d", id)
```
Use "http/%d" for the channel?

ry commented on http.go:
```diff
@@ -0,0 +1,112 @@
+	s.Addr = listenAddr
+	s.Handler = http.HandlerFunc(handler)
+	go s.ListenAndServe()
+
```
empty line (run go fmt?)

ry commented on msg.proto:
```diff
@@ -99,4 +107,22 @@ message Msg {
+  int32 http_res_id = 139;
+  bytes http_res_body = 140;
+  int32 http_res_code = 141;
+  repeated string http_res_headers = 142;
```
In the future, I'm hoping to reuse the structure for FETCH_RES for server messages (by adding a new message type) . But no changes requested now.

matiasinsaurralde replied:
Sounds fine

ry commented on runtime.ts:
```diff
@@ -314,6 +317,8 @@ class TypeScriptHost implements ts.LanguageServiceHost {
       } else if (name === "typescript") {
         resolvedFileName = resolveModuleName("typescript.d.ts", "/$asset$/");
+      } else if (name === "http") {
+        resolvedFileName = resolveModuleName("http.d.ts", "/$asset$/");
```
Can you merge all the http stuff into the Deno namespace? Maybe it can be split up later, but for now I prefer to keep it flat.

matiasinsaurralde replied:
Sure

ry commented on http.d.ts:
```diff
@@ -0,0 +1,44 @@
+import { main as pb } from "./msg.pb";
```
Move this stuff into deno.d.ts

matiasinsaurralde commented:
@ry made all the tweaks:
- Merged http definitions into the deno namespace.
- Simplified the reqID usage.
- Removed the unused options field from Response.
- Requests now use `http/[request -id]` as channel name.
- Lint.
I will add tests in the next hours!

peterwmwong commented on http.ts:
```diff
@@ -0,0 +1,127 @@
+        return res;
+    }
+    onMsg(msg: pb.Msg) {
+        if (msg.command = pb.Msg.Command.HTTP_REQ) {
```
Is this suppose to be `===`?

matiasinsaurralde replied:
@peterwmwong Fixed, thanks

matiasinsaurralde commented:
@ry Great, added tests now!

yordis commented on http.ts:
```diff
@@ -0,0 +1,129 @@
+    // tslint:disable-next-line: no-any
+    body: any | string;
+    constructor(url: string, opts?: RequestOptions) {
+      if (opts == null) {
```
@matiasinsaurralde I think you meant `undefined`.

`?` in Typescript is not `null` but `undefined`.

ry replied:
`x == null` is often used to testing against both null and undefined. 
tslint makes exceptions for triple equals in this case.

I think it's ok.

yordis replied:
@ry fair enough that makes sense I just prefer to the code to mirror what the type system says

Related to https://github.com/ry/deno/issues/234#issuecomment-396042429

ry commented on msg.proto:
```diff
@@ -99,4 +107,22 @@ message Msg {
   // write_file_sync_perm specified by https://godoc.org/os#FileMode
+
+  // HTTP_CREATE
+  int32 http_server_id = 133;
```
For each command type, the prefix of the members should contain the command name. So this should be called `http_create_server_id`. I want to strictly enforce this so keep this sane as it grows.

Furthermore, you should start the indexing at 10 times the index of the command. That is, since `HTTP_CREATE = 14`, this should start at 140.
```
int32 http_create_server_id = 140;
```

Finally, I think HTTP_CREATE is unnecessary. Why not have HTTP_LISTEN response with the id?

ry commented on msg.proto:
```diff
@@ -99,4 +107,22 @@ message Msg {
+  int32 http_server_id = 133;
+
+  // HTTP_LISTEN
+  int32 http_listen_port = 134;
```
Index should start at 150

ry commented on msg.proto:
```diff
@@ -99,4 +107,22 @@ message Msg {
+  int32 http_listen_port = 134;
+
+  // HTTP_REQ
+  int32 http_req_id = 135;
```
Index should start at 160

ry commented on msg.proto:
```diff
@@ -99,4 +107,22 @@ message Msg {
+  bytes http_req_body = 138;
+
+  // HTTP_RES
+  int32 http_res_id = 139;
```
Index should start at 170

ry commented on msg.proto:
```diff
@@ -25,6 +25,14 @@ message Msg {
+    HTTP_RES = 17;
+    HTTP_RES_WRITE = 18;
+    HTTP_RES_END = 19;
+    HTTP_RES_STATUS = 20;
```
Rename HTTP_RES_STATUS to HTTP_RES_HEADER

ry commented on msg.proto:
```diff
@@ -99,4 +107,22 @@ message Msg {
+  bytes http_res_body = 140;
+  int32 http_res_code = 141;
+  repeated string http_res_headers = 142;
 }
```
Please add HTTP_RES_STATUS, HTTP_RES_WRITE, HTTP_RES_END here. And do not reuse fields from other commands.

ry commented on http.ts:
```diff
@@ -0,0 +1,129 @@
+    status(code: number) {
+        pubInternal(this.channel, {
+            command: pb.Msg.Command.HTTP_RES_STATUS,
+            httpResCode: code
```
And do not reuse fields from other commands.

s/httpResCode/httpResHeaderCode/

ry commented on http.ts:
```diff
@@ -0,0 +1,129 @@
+    }
+    status(code: number) {
+        pubInternal(this.channel, {
+            command: pb.Msg.Command.HTTP_RES_STATUS,
```
s/HTTP_RES_STATUS/HTTP_RES_HEADER/

ry commented on http.ts:
```diff
@@ -0,0 +1,129 @@
+        }
+        pubInternal(this.channel, {
+            command: pb.Msg.Command.HTTP_RES_WRITE,
+            httpResBody: rawData
```
Do not reuse fields from other commands. s/httpResBody/httpResWriteBody/

ry commented on http.ts:
```diff
@@ -0,0 +1,129 @@
+        servers.set(this.id, this);
+        pubInternal("http", {
+            command: pb.Msg.Command.HTTP_CREATE,
+            httpServerId: this.id
```
And do not reuse fields from other commands: s/httpServerId/httpCreateServerId/

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ry/deno?pullRequest=209) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ry/deno?pullRequest=209) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: matiasinsaurralde<br/>:x: ry<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ry/deno?pullRequest=209) it.</sub>

matiasinsaurralde commented:
@ry I've re-organized the Protobuf commands, added the `wg` interactions and implemented `server.close()`.

matiasinsaurralde commented:
Oh it seems the Go version running in the CI doesn't support `http.Close`, should we update? I'm running Go 1.10.

ry commented:
@matiasinsaurralde LGTM!

I have one more set of small adjustments and I'm trying to fix travis in this branch: https://github.com/ry/deno/tree/http-server-tweaks

Once that's working, I will squash and land

matiasinsaurralde commented:
@ry Cool, thanks!

alexhultman commented:
So if I understand this thread correctly,

We are applauding a prototype, rewritten server, doing significantly worse in a hello world throughput test than even Node.js?

Do I need to remind you Node.js already performs among the worst of all servers for this case? We're talking 10x here. I do 6x Node.js _with SSL_ so this is nothing but straight out horrible.

If anything this proves how Deno _NOT_ should do it.

ry commented:
@matiasinsaurralde Sorry about this but I'm going to close this PR. It seems to have inspired your work on rust-v8worker2 ! So I'm we can merge efforts in a few days once we've got the runtime running : )
Many thanks for your patience.
